### ISSUE TITLE
Backporting of the FMI 3.0 Terminals and Icons approach to FMI 2.0.4

### ISSUE THREAD

**Issue by TorstenBlochwitz** [modelica/fmi-standard#1799]
The FMI 2.0.4. specification simply points to the respective sections within the FMI 3.0 specification.

From my point of view, a clarification on how to reference non-scalar variables as TerminalMemberVariables is needed. Contrary to the FMI 3.0 specification (which supports non-scalar variables directly) in FMI 2.0 a special naming convention must be used.

Example:
- modelDescription.xml contains a vector "x" of size 3, expressed via the 3 scalar variables "x[1]", "x[2]" and "x[3]" following the structured naming convention.
- Is it allowed to use "x" as TerminalMemberVariable, which then points to the 3 variable above? Or do we have to introduce 3 TerminalMemberVariables "x[1]", "x[2]" and "x[3]"?

**Comment by chrbertsch** [comment]
> * Is it allowed to use "x" as TerminalMemberVariable, which then points to the 3 variable above? Or do we have to introduce 3 TerminalMemberVariables "x[1]", "x[2]" and "x[3]"?

In the first case we would have to introduce new a new rule how to interpret `variableName `of a `TerminalMemberVariable` and how to apply getXXX and setXXX calls, while in the second case a non-normative sentence in the FMI 2.0.4 spec would be sufficient, as we just reference already existing FMI 2.0 variables. 
Thus I would prefer the second proposal if there are no strong arguments for the first.

**Comment by TorstenBlochwitz** [comment]
I would highly prefer the first variant. Why do we have to introduce a new rule on how to apply fmi2getXXX and fmi2setXXX? If we introduce a rule like "If a TerminalMemberVariable points to variables which are defined as nonscalar variables using the structured naming convention, all elements of this moredimensional variables is meant.

By the way: in FMI 3.0 we still have the structured naming convention and non-scalar variable could be expresed using it. Hence it might be a good idea to add such a rule to FMI 3.0 too.

**Comment by chrbertsch** [comment]
> By the way: in FMI 3.0 we still have the structured naming convention and non-scalar variable could be expresed using it. Hence it might be a good idea to add such a rule to FMI 3.0 too.

If I see it right, FMI 3 does not forbid to have variables `x`, `x[1]`, where `x[1]` is a separate variable and not the first array element of `x`. Ugly, but allowed, right?
Then with you proposal, `x` in the Terminals definition would not be unique, and we would need more clarifying rules.

If users want the full benefit of arrays, they should use FMI3 :-)

**Comment by CSchulzeTLK** [comment]
The FMI 2 and [3 specifications](https://fmi-standard.org/docs/3.0/#namingSection) explicitly mention `arrayIndices` in the "precise syntax" of the structured naming convention, as both copied text from the Modelica specification. This seems to be normative text.

![grafik](https://user-images.githubusercontent.com/33543611/187532294-27c05c2a-40af-481c-bef7-0b7870dc744d.png)

The specification doesn't explicitly forbid having `x` and `x[1]`. Even ugly incomplete arrays are allowed:
"It might be that not all elements of an array are present. If they are present, they are given in consecutive order in the XML file."
I think the `x`-`x[1]` issue needs to be fixed.

So, IMHO it is defined how arrays are represented in the structured naming convention. Referencing an array by giving the array name instead of all scalar members is technically possible.

However, my interpretation of the current specifications 2/3 is, that this is not allowed. All terminal members are identified by the name:
"The normalized string variableName identifies a variable in <ModelVariables>"
IMHO "identifies" means that the strings have to be equal. At least that is what I expected when we were discussing this. But some people might have a wider interpretation of the word "identifies".

**Comment by pmai** [comment]
> "The normalized string variableName identifies a variable in < ModelVariables >"
> IMHO "identifies" means that the strings have to be equal. At least that is what I expected when we were discussing this. But some people might have a wider interpretation of the word "identifies".

Even the most wide interpretation of "identifies" would need to, well, uniquely identify a variable. As having both x (either scalar or array) and x[1] is allowed in both FMI 2.0 and 3.0, allowing "x" to match x[1] would make such references no longer unambiguous.

This seems neither wise nor unproblematic.

**Comment by CSchulzeTLK** [comment]
In fact the idea would be to not uniquely identify a variable but rather all scalar variables of an array.

The question is, if having `x` and `x[1]` has been allowed up to now. If the structured naming convention allows variables to end with the array index in squared brackets, then the variable name before the array must be the name of the array. I think this rather is a clarification than an actual change.

@klausschuch, Is the derivative attribute properly defined for arrays? Is it possible to have an FMI 3 FMU with an array model variable (e.g. "x"), and provide single derivatives of it as scalar variables (e.g. "der(x[2])"). Variables having the same derivative attribute seems to be allowed. So several scalar derivatives could reference the same array model variable.
At least Modelica tools might provide only a subset of derivatives of an array.

**Comment by chrbertsch** [comment]
Design meeting: 
Klaus: There was a similar discussion in SSP, whether  a connection of x and y is possible, if only x[1], x[2] .... and y[1], y[2] are defined. There  we decided to use explicit names x[1]
Torsten S: We should consider whether indexing is one or zero based indexing
Christian S: You could not connect termals of an FMI 2.0 FMU with x[1], x[2] with an FMI 3.0 FMU with an array x
Andreas: I would need to see an example. There can be holes in the definition of the structured naming convention, e.g. x[1], x[3] but no x[2]. We cannot make assumptions. In Modelica there can be strange "arrays" with different units etc.
Klaus: If this problem occurs, the FMU is badly exported.
Christian S: Do we have the same problem with derivatives in FMI 3.0?
Klaus: no, you can only reference the whole array. The derivative reference references a value reference, i.e.,  a whole array.
Christian S: I can make an example with fluid ports.
Andreas: In the example we would need an example to connect an 2.0 and a 3.0 with arrays.
Klaus: if we allow references to "x" in FMI 2.0, in FMI 3.0 we would have to forbid it.
Andreas: If we allow "meta" variable x, we would have to add additional assumptions
Christian S: in FMI 3.0 it is only stated that the name "identifies" a variable
Klaus, Andreas: "variable" is exactly defined.
Christian: Could we add additional checks if such "incompatible" terminals can be connected
Andreas: This adds complexity beyond the interface description
Klaus: I think "meta" variables in FMI 3.0 could be used, if described well.
Christian S: in FMI 3.0 we have the same problem
Klaus: No, in FMI 3.0 it is not allows.
Andreas: if we do not fix this, we will have a problem with the interoperability of FMI2.0 and FMI3.0 FMUs.
Torsten S: you could wrap an FMI 2.0 to an FMI 3.0 FMU as a workaround
Andreas: There could be a problem with the state machines
Christian S: There are slight incompabilities.
Klaus: We can define, how FMI 2.0 variables should look like that they ccould be used  as a meta variable. (Consecutive order of varables, if x and x[....] are not both present -->  the name of the meta variable cannot be used already)
Torsten: We would have to consider attribus, types, units ....
Klaus: We could say: all attributes must be exaclty the same
Torsten: Think about everything ... e.g. derivatives
Torsten: it would be great to have a reference FMU for this.

Summary: let's try out this "meta"/"rolled"/"meta array"/"virtual array"/"naming convention array"/"implicit" variable approach
--> @CSchulzeTLK will create a proposal until Friday

**Comment by pmai** [comment]
> In fact the idea would be to not uniquely identify a variable but rather all scalar variables of an array.

While this is the idea (which has its own problems, as likely discussed today), it is beside the point, due to the x vs x[...] issue:

> The question is, if having `x` and `x[1]` has been allowed up to now. 

There is no question that it is allowed in both 2.0 and 3.0 (there being no normative restrictions on the variable names that would make this disallowed) to have distinct variables named x and e.g. x[1] at the same time. Hence not only do terminals have to refer to the full name since that is the current wording, changing this would make for ambiguous references (and no way out).

So for FMI 3.0 this cannot be changed, as it would make currently valid FMUs invalid and/or change their meaning. For 2.0 we could add additional restrictions that apply only to FMUs that use terminals, but that would make the FMI 2 and 3 mechanisms different for no appreciable benefit.

So currently I'm wondering from my sunny beach why this is even being persued, especially since this complex feature adds no functionality, and only saves on some XML...

> I think this rather is a clarification than an actual change.

I cannot see it this way...

**Comment by chrbertsch** [comment]
> , but that would make the FMI 2 and 3 mechanisms different for no appreciable benefit.

Without a "naming convention array", important use cases will not be possible, like the creation of compatible terminals for FMI 2.0. and 3.0 FMUs from fluid systems with mass fractions arrays.

**Comment by pmai** [comment]
I'm unsure of the exact meaning of the whole "meta variable" wording in today's discussion, but it seems to me only tangentially related to the terminal discussion: There are obvious approaches to connecting FMI 2.0/3.0 structured naming convention arrays to FMI 3.0 native arrays, with the usual limitations due to the loose 2.0/3.0 specs. These carry over exactly to terminals, without any needs or benefits from more complex terminal rules (given that 2.0 currently does not have terminals, a solution must work for non-terminal stuff if it is to be relevant).

SSP probably has more needs to more intelligently support these kind of connections, and even there it is not clear that more complex rules in FMI 2.0 terminals would help rather than hinder...

**Comment by pmai** [comment]
> > , but that would make the FMI 2 and 3 mechanisms different for no appreciable benefit.
> 
> 
> 
> Without a "naming convention array", important use cases will not be possible, like the creation of compatible terminals for FMI 2.0. and 3.0 FMUs from fluid systems with mass fractions arrays.

I do not see how this is the case:

The terminal definitions are per FMU. The 2.0 one will include all the scalar variables that make up the naming convention array. The 3.0 one, assuming it uses native arrays, will include the array variable. Compatibility is checked based on the structured naming convention rules (and whatever additional compatibility rules underly the terminal kind, etc.).

This works for terminals the same way it works for non-terminal connections between structured and native arrays (and why should it differ, since terminals are not about arrays at all).

Inventing new rules of how variables are referenced in the terminal definition does not make any of this easier, in fact it makes it harder (since now resolving the terminal to its member variables has become more involved).

If we want to do anything here, we should point out the obvious and not so obvious ways to resolve connections between structured naming arrays and native arrays in the IG, and clarify in 3.0 and 2.0 terminal definitions that connection resolution using names in terminals is not meant to preclude this kind of mapping.

**Comment by CSchulzeTLK** [comment]
[2022-09-01 FMI 3 Terminals.zip](https://github.com/modelica/fmi-standard/files/9469528/2022-09-01.FMI.3.Terminals.zip)

I hope this is what we were discussing. I prepared 4 FMUs:
1. BoundaryExample_fmi3_scalarVariable_scalarTerminalMember.fmu
2. BoundaryExample_fmi3_arrayVariable_arrayTerminalMember.fmu
3. BoundaryExample_fmi2_scalarVariable_scalarTerminalMember.fmu
4. BoundaryExample_fmi2_scalarVariable_arrayTerminalMember.fmu

(1) and (3) should be compatible to each other, and (2) and (4) should be compatible to each other.

Due to the matchingRule "plug" matching of the member variables is done based on the memberName. So a member which is an array is not matched to several scalar member variables in another terminal.

The 2.0 FMU has scalar variables:
```XML

    <!-- Index for next variable = 4 -->
    <ScalarVariable
      name="ports[1].Xi_outflow[1]"
      valueReference="67108866"
      description="Independent mixture mass fractions m_i/m close to the connection point if m_flow &lt; 0"
      causality="output">
      <Real
        quantity="MassFraction"
        unit="kg/kg"
        min="0.0"
        max="1.0"
        nominal="0.1"/>
    </ScalarVariable>
    <!-- Index for next variable = 5 -->
    <ScalarVariable
      name="ports[1].Xi_outflow[2]"
      valueReference="67108867"
      description="Independent mixture mass fractions m_i/m close to the connection point if m_flow &lt; 0"
      causality="output">
      <Real
        quantity="MassFraction"
        unit="kg/kg"
        min="0.0"
        max="1.0"
        nominal="0.1"/>
    </ScalarVariable>
```

... and a terminal either uses the array name "ports[1].Xi_outflow":
```XML
  <Terminals>
    <Terminal>
      <Terminal
        name="ports[1]"
        matchingRule="plug"
        terminalKind="Modelica.Fluid.Interfaces.FluidPorts_b">
        <TerminalMemberVariable
          variableName="ports[1].p"
          memberName="p"
          variableKind="signal"/>
        <TerminalMemberVariable
          variableName="ports[1].m_flow"
          memberName="m_flow"
          variableKind="inflow"/>
        <TerminalStreamMemberVariable
          inStreamVariableName="ports[1].h_outflow_inStream"
          inStreamMemberName="h_outflow_inStream"
          outStreamVariableName="ports[1].h_outflow"
          outStreamMemberName="h_outflow"/>
        <TerminalStreamMemberVariable
          inStreamVariableName="ports[1].Xi_outflow_inStream"
          inStreamMemberName="Xi_outflow_inStream"
          outStreamVariableName="ports[1].Xi_outflow"
          outStreamMemberName="Xi_outflow"/>
      </Terminal>
    </Terminal>
  </Terminals>
```

... or the terminal uses all scalar names "ports[1].Xi_outflow[*]":

```XML
  <Terminals>
    <Terminal>
      <Terminal
        name="ports[1]"
        matchingRule="plug"
        terminalKind="Modelica.Fluid.Interfaces.FluidPorts_b">
        <TerminalMemberVariable
          variableName="ports[1].p"
          memberName="p"
          variableKind="signal"/>
        <TerminalMemberVariable
          variableName="ports[1].m_flow"
          memberName="m_flow"
          variableKind="inflow"/>
        <TerminalStreamMemberVariable
          inStreamVariableName="ports[1].h_outflow_inStream"
          inStreamMemberName="h_outflow_inStream"
          outStreamVariableName="ports[1].h_outflow"
          outStreamMemberName="h_outflow"/>
        <TerminalStreamMemberVariable
          inStreamVariableName="ports[1].h_outflow_inStream"
          inStreamMemberName="h_outflow_inStream"
          outStreamVariableName="ports[1].h_outflow"
          outStreamMemberName="h_outflow"/>
        <TerminalStreamMemberVariable
          inStreamVariableName="ports[1].Xi_outflow_inStream[1]"
          inStreamMemberName="Xi_outflow_inStream[1]"
          outStreamVariableName="ports[1].Xi_outflow[1]"
          outStreamMemberName="Xi_outflow[1]"/>
        <TerminalStreamMemberVariable
          inStreamVariableName="ports[1].Xi_outflow_inStream[2]"
          inStreamMemberName="Xi_outflow_inStream[2]"
          outStreamVariableName="ports[1].Xi_outflow[2]"
          outStreamMemberName="Xi_outflow[2]"/>
        <TerminalStreamMemberVariable
          inStreamVariableName="ports[1].Xi_outflow_inStream[3]"
          inStreamMemberName="Xi_outflow_inStream[3]"
          outStreamVariableName="ports[1].Xi_outflow[3]"
          outStreamMemberName="Xi_outflow[3]"/>
        <TerminalStreamMemberVariable
          inStreamVariableName="ports[1].Xi_outflow_inStream[4]"
          inStreamMemberName="Xi_outflow_inStream[4]"
          outStreamVariableName="ports[1].Xi_outflow[4]"
          outStreamMemberName="Xi_outflow[4]"/>
      </Terminal>
    </Terminal>
  </Terminals>
```

**Comment by chrbertsch** [comment]
Looking at Christian S's examples and Pierres comments: 
**could you please comment and state your opinion?**

(this is the last open point for the release of FM 2.0.4. Once we have more feedback, we can decide if we need another meeting to discuss and decide this.)

My personal opinion: Following Pierre's arguments, I would stay in FMI 2.0.x with referencing the scalarized array elements in the terminals, and let the importing tools do the matching logic. This could be explained in the IG, and clarifying comments could be added to FMI 2.0.4 and FMI 3.0.1

**Comment by andreas-junghanns** [comment]
Thank you @CSchulzeTLK for the examples!
They show exactly what we should clearly define.
E.g. While the 3.0 FMU references "ports[1].Xi_outflow" and the importer will find that in the variables of the modelDescription.xml,
the 2.0.4 FMU also references "ports[1].Xi_outflow" in its Terminal and our text must explain that this means the virtual variable array composed of "ports[1].Xi_outflow[1]" up until "ports[1].Xi_outflow[4]".
We could introduce a different notation, like "ports[1].Xi_outflow[]" or "ports[1].Xi_outflow[1..4]" to be more precise and help the importer in the matching.

@chrbertsch : I have not seen a PR for this discussion. Is that not available (yet)?

**Comment by CSchulzeTLK** [comment]
Compatibility of Arrays in Terminals
-------------------------------------

In FMI 3 an exporting tool may either choose to export an array as a single variable in `ModelVariable` (e.g. "x"), or flatten the array to the scalar elements (e.g. "x[1], x[2], ...").
Both, an array `ModelVariable` and a scalar array element `ModelVariable` may be used as `variableName`/`inStreamVariableName`/`outStreamVariableName` in a `TerminalMemberVariable`/`TerminalStreamMemberVariable`. 

The `matchingRule` "plug" and "bus" use the `memberName`/`inStreamMemberName`/`outStreamMemberName` for variable matching in connected terminals.
A terminal member variable is matched to just one terminal member variable in another terminal. So an array variable in a terminal is not mapped to several terminal member variables which refer to the flattened scalar array elements.
_[I.e. a terminal with the array ['x'] is not compatible with a terminal with ['x[1]', 'x[2]', 'x[3]']]_

An array `ModelVariable` in a terminal with `matchingRule` "sequence" is inserted in the terminal sequence in the serialized form described in section [...serialization c format...].





Virtual Array Names
---------------------

FMI 2 does not support array variables. Arrays are flattened to a list of scalar elements. A terminal may reference all scalar members of an array in a 2.0 FMU (contrary to a 3.x FMU) with the virtual array name as `variableName`/`inStreamVariableName`/`outStreamVariableName` of the `TerminalMemberVariable`/`TerminalStreamMemberVariable`. The array name is virtual, as there is no scalar variable having this variable name, but several scalar variables having the virtual name with the appended array indices in squared brackets.

The following requirements have to be fulfilled, to reference an array with its virtual array name:
- The FMU uses the structured naming convention.
- The scalar array member variables end with the array indices in squared brackets.
	The syntax is "B-name [ arrayIndices ] ", whereas "B-name" is the virtual array name (see the precise syntax in the structured naming convention).
- The array is 0 or 1 based.
- All elements of the array must be present.
- A ScalarVariable with the virtual array name must not exist.
	[I.e. if ScalarVariable 'x[1]' exists then ScalarVariable 'x' must not exist).]

[
shortened example...	
]

**Comment by pmai** [comment]
I still do not see any benefit to this approach, which will by necessity make 2.0 and 3.0 terminal definitions differ (since 3.0 is already set in stone), not solve the same problem that still exists for 3.0 structured naming conventions, and not address connecting 2.0 FMUs without terminals to 3.0 terminals. [Disregarding the known problem of ambiguous scalar variable names].

What is wrong with just clarifying that terminal connections (like all connections) can take structured naming into account when connecting native arrays and structured naming arrays even when using terminals?

**Comment by CSchulzeTLK** [comment]
I also prefer to have the same rules for FMI 2 and 3. I would prefer a solution that enables users of FMI 3 to connect terminals with flattened arrays to terminals with actual arrays. Not all tools might implement the FMI 3 arrays right away.

>What is wrong with just clarifying that terminal connections (like all connections) can take structured naming into account when connecting native arrays and structured naming arrays even when using terminals?

I agree, this could also be an option. But consider the specification text about `matchingRule` "plug":
>Matching of the variables is based on memberName. An importer should connect terminals **only if all member variables are present and match**.
(https://fmi-standard.org/docs/3.0/#table-predefined-matching-rules)

If we agree that a clarification is sufficient, this is fine by me. This means we shift the resposibility to the importer.

Nevertheless, we might need to add the bullet points above as non-normative text, and clarify the array serialization for matchingRule "sequence".

**Comment by pmai** [comment]
> >What is wrong with just clarifying that terminal connections (like all connections) can take structured naming into account when connecting native arrays and structured naming arrays even when using terminals?
> 
> 
> 
> I agree, this could also be an option. But consider the specification text about `matchingRule` "plug":
> 
> >Matching of the variables is based on memberName. An importer should connect terminals **only if all member variables are present and match**.
> 
> (https://fmi-standard.org/docs/3.0/#table-predefined-matching-rules)
> 
> 
> 
> If we agree that a clarification is sufficient, this is fine by me. This means we shift the resposibility to the importer.
> 

Yes, that would be the approach I had in mind: Add a clarification that "match" includes matching based on structured naming conventions besides pure name-equality (which I think is only a clarification since we never specified the exact meaning of match in this context as far as I'm aware).

> Nevertheless, we might need to add the bullet points above as non-normative text, and clarify the array serialization for matchingRule "sequence".

Agreed, we should give some more guidance here and add further guidance to the implementers guide.

**Comment by chrbertsch** [comment]
Design Webmeeting: 

If we clarify "matching" in 3.0.1, in 2.0.4 we do not have to change anything, except perhaps changing the link to FMI 3.0 to "3.0 or latest 3.0.x maintenance release".

Antoine: I would reference an exact 3.0.1 version.
Andreas: this creates a larger maintenance effort. Let's reference "3.0"

-> current proposal for FMI 2.0.4 is sufficient.

**Comment by chrbertsch** [comment]
Closing this issue with milestone 2.0.4, as it was decided that nothing has to be done for 2.0.4.
Fix for 3.0.1 will be in PR #1808
